zallet/src/components/json_rpc/methods/view_transaction.rs

```diff
@@ -0,0 +1,480 @@
+            ShieldedProtocol::Orchard => ("orchard", "action"),
+        };
+
+        wallet
+            .with_raw(|conn| {
+                conn.query_row(
+                    &format!(
+                        "SELECT txid, {output_prefix}_index, address, value
+                        FROM {pool_prefix}_received_notes
+                        JOIN transactions ON tx = id_tx
+                        JOIN addresses ON address_id = addresses.id
+                        WHERE nf = :nf"
+                    ),
+                    named_params! {
+                        ":nf": nf,
+                    },
+                    |row| {
+                        Ok((
+                            TxId::from_bytes(row.get("txid")?),
+                            row.get("output_index")?,
+                            row.get("address")?,
+                            Zatoshis::const_from_u64(row.get("value")?),
+                        ))
+                    },
+                )
+            })
+            .optional()
+            .map_err(|e| {
+                LegacyCode::Database.with_message(format!("Failed to fetch spent note: {:?}", e))
+            })
+    }
+
+    /// Fetches the address that was cached at transaction construction as the recipient.
+    // TODO: Move this into `WalletRead`.
+    fn sent_to_address(
+        wallet: &DbConnection,
+        txid: &TxId,
+        pool: ShieldedProtocol,
+        idx: u16,
+        fallback_addr: impl FnOnce() -> Option<String>,
+    ) -> RpcResult<Option<String>> {
+        Ok(wallet
+            .with_raw(|conn| {
+                conn.query_row(
+                    "SELECT to_address
+                            FROM sent_notes
+                            JOIN transactions ON tx = id_tx
+                            WHERE txid = :txid
+                            AND   output_pool = :output_pool
+                            AND   output_index = :output_index",
+                    named_params! {
+                        ":txid": txid.as_ref(),
+                        ":output_pool": match pool {
+                            ShieldedProtocol::Sapling => 2,
+                            ShieldedProtocol::Orchard => 3,
+                        },
+                        ":output_index": idx,
+                    },
+                    |row| row.get("to_address"),
+                )
+            })
+            // Allow the `sent_notes` table to not be populated.
+            .optional()
+            .map_err(|e| {
+                LegacyCode::Database.with_message(format!("Failed to fetch sent-to address: {}", e))
+            })?
+            // If we don't have a cached recipient, fall back on an address that
+            // corresponds to the actual receiver.
+            .unwrap_or_else(fallback_addr))
+    }
+
+    if let Some(bundle) = tx.sapling_bundle() {
+        let incoming: BTreeMap<u16, (sapling::Note, Option<sapling::PaymentAddress>, [u8; 512])> =
```

Comment:
Why do we need to re-do trial decryption? Don't we store sufficient information in the `*_received_notes` tables?

Reply:
I implemented this the same way as `zcashd` does to ensure identical semantics. Now it happens that `zcashd` decrypts _everything_ on-demand because it stores the shielded information once non-relationally (in `CWalletTx`). So we may be able to switch to the data stored in the various tables, as a performance optimisation. But we would need to determine whether the _semantics_ of the information in those tables is consistent with what `z_viewtransaction` provides, which would take longer.